etl/steps/data/garden/who/2022-09-30/ghe.py

```diff
@@ -139,19 +136,137 @@ def build_custom_age_groups(df: pd.DataFrame) -> pd.DataFrame:
-    df_sh["death_rate100k"] = 100000 * df_sh["death_count"] / df_sh["population"]
-    df_sh["flag_level"] = 3
-    df_sh = df_sh.drop(columns=["population"])
-
-    log.info("ghe: concatenate dfs")
-    df = pd.concat([df, df_sh], axis=0, ignore_index=True, sort=False)
+    df = add_population(
+        df=df,
+        country_col="country",
+        year_col="year",
+        sex_col="sex",
+        sex_group_all="Both sexes",
+        sex_group_female="Female",
+        sex_group_male="Male",
+        age_col="age_group",
+        age_group_mapping=AGE_GROUPS_RANGES,
+    )
+    assert df["population"].isna().sum() == 0
+    df_age_group_ihme = build_custom_age_groups(df, age_groups=age_groups_ihme)
+    df_age_group_self_harm = build_custom_age_groups(df, age_groups=age_groups_self_harm, select_causes=["Self-harm"])
+    df = remove_granular_age_groups(df, age_groups_to_keep=["ALLAges"])
+    df_combined = pd.concat([df, df_age_group_ihme, df_age_group_self_harm], axis=0)
+    df_combined = df_combined.loc[:, ~df_combined.columns.duplicated()]
+
+    return df_combined
+
+
+def combine_drug_and_alcohol(df: pd.DataFrame) -> pd.DataFrame:
+    substance_use_disorders = ["Drug use disorders", "Alcohol use disorders"]
+    substance_df = df[df["cause"].isin(substance_use_disorders)]
+    substance_agg = (
+        substance_df.groupby(["country", "year", "age_group", "sex"], as_index=False, observed=True)[
+            ["daly_count", "death_count"]
+        ]
+        .sum()
+        .reset_index()
+    )
+    substance_agg = calculate_rates(substance_agg)
+    return df
+
+
+def calculate_rates(df: pd.DataFrame) -> pd.DataFrame:
+    if all(df.columns != "population"):
+        df = add_population(
+            df=df,
+            country_col="country",
+            year_col="year",
+            sex_col="sex",
+            sex_group_all="Both sexes",
+            sex_group_female="Female",
+            sex_group_male="Male",
+            age_col="age_group",
+            age_group_mapping=AGE_GROUPS_RANGES,
+        )
+    df["daly_rate100k"] = 100000 * (df["daly_count"] / df["population"])
+    df["death_rate100k"] = 100000 * (df["death_count"] / df["population"])
+    df["daly_rate100k"] = np.where(df["daly_count"] == 0, 0, df["daly_rate100k"])
+    df["death_rate100k"] = np.where(df["death_count"] == 0, 0, df["death_rate100k"])
+    df = df.drop(columns=["population"])
+
+    return df
+
+
+def build_custom_age_groups(df_age: pd.DataFrame, age_groups: dict, select_causes: Any = None) -> pd.DataFrame:
+    """
+    Estimate metrics for broader age groups. In line with the age-groups we define in the age_groups dict:
+    """
+    # Add population values for each dimension
+    age_groups_to_drop = (
+        df_age["age_group"].drop_duplicates()[~df_age["age_group"].drop_duplicates().isin(age_groups.keys())].to_list()
+    )
+    log.info(f"Not including... {age_groups_to_drop} in custom age groups")
```

Comment:
I am not sure I understand what is going on here, but perhaps it is equivalent to the following?

```python
age_groups_to_drop = set(df_age["age_group"]) - set(age_groups)
log.info(f"Not including... {age_groups_to_drop} in custom age groups")
```

I think this is slightly easier to read

Reply:
Thanks!